crates/matrix-sdk/src/sliding_sync/mod.rs

```diff
@@ -1069,41 +1084,54 @@ impl SlidingSync {
         //
         // That's why we are running the handling of the response in a blocking
         // mode since it cannot be cancelled abruptly.
-        debug!("Sliding sync response received");
+        let this = self.clone();
+        let stream_id = stream_id.to_owned();
 
-        match &response.txn_id {
-            None => {
-                error!(stream_id, "Sliding Sync has received an unexpected response: `txn_id` must match `stream_id`; it's missing");
-            }
+        // Spawn a new task to ensure that the code of this task cannot be
+        // cancelled if this method is cancelled.
+        spawn(async move {
+            debug!("Sliding sync response received");
 
-            Some(txn_id) if txn_id != stream_id => {
-                error!(
+            // Here we go. We lock the `response_handling_lock` to ensure that
+            // only one response can be handled at a time.
+            let global_lock = this.response_handling_lock.lock().await;
```

Comment:
Why not lock `list_generators` for the same duration instead? It would have the same effect.

Reply:
The `list_generators` lock is created inside `SlidingStream::stream`, whilst `response_handling_lock` belongs to `SlidingSync`, and actually to all `SlidingSync` as it's shared between all the clones.